### DISCUSSION TITLE
Potential PSyclone projects

### DISCUSSION THREAD

**Discussion by arporter** [stfc/PSyclone#3329] [Ideas]
This Issue is a place for us to capture ideas for potential PSyclone-related projects. The idea is that they be self-contained, fairly short-term pieces of work that might be suitable for a student to take on.

1. A PSyIR-aware plugin for e.g. emacs or vim
2. A GUI that allows a user to manipulate PSyIR (or even generate from scratch) c.f. Scratch/Blockly

**Comment by hiker** [comment]
VERY rough brainstorming only - just some key ideas, many of the ideas need some thorough fleshing out (and maybe the idea is even stupid in the first place :) )
1. PSyData Applications (some of which I am interested in)
    - in-situ visualisation for LFRic (probably done by Wolfgang) and Gocean/Nemo?
    - Kernel Extraction for nemo(?) 
    - sensibility analysis: use kernel extraction to determine how small changes in a parameter affect the result (to help detecting what might create an numerical instability)
    - Other profiling wrapper (score-p, Cray-pat, ...)
2. Can PSyclone be combined with python? This is based on a suggestion I had from R&D - the idea was compared with the way machine learning libraries work: you specify something in python, but the actual maths is done on GPUs etc. (_this idea needs a lot of work to define what exactly is required_)
    - The idea is that the main program and a lot of non-performance critical code is written in python, but the performance critical code is done with PSyclone+Fortran.
    - python would need to be able to call infrastructure library (e.g. to create fields), and invoke-subroutines
    - maybe kernels could be called in Fortran to allow for easier prototyping??? Yes, that opens a whole world of pain I guess - but might be doable with a python psyir backend???
3. Comparable debugging: couple two versions of one applications ('orig' and 'new'), which would automatically compare results after each kernel. This could be useful to e.g. verify an optimised or threaded version of a kernel, or the use of a different compiler. After each kernel statistics could be collected about the result differences in one kernel. Then the 'orig' results would be transferred to the 'new' kernel and used further on (instead of the results computed in 'new' itself), i.e. each kernel would be called with identical input parameter. 
    - At the end we might be able to deduce statistics about which functions might be wrong (result differences outside a certain epsilon)
    - which functions  contribute how much to result differences
    - identify kernels that introduce a dependency on the domain decomposition used?
4. Can PSyclone be used to tackle the 'correctness' problem? I.e. to determine if results obtained by using a new/different compiler are still 'correct'? Maybe PSyclone could create kernels that determine an expected interval in which the output values are expected to be?
    - A simple approach could be to take an input parameter I, and call the kernel with I*(1-x) and then with I*(1+x). This can be done for each input parameter I1, .. In, resulting in all in all 2^n kernel invocations - which might tell is in which range the output values are expected to be. If then the kernel compiled with the new/different compiler called with I1, ... In (ie. unmodified) is outside the range determined above, chances are there is a code or compiler bug??
5. Creating unit tests for kernels? That might be reasonable easy with psyke(?)
6. Kernel verification: e.g. determine that field accesses are within the specified stencil (either psyclone-time analysis, or runtime if required(??)), read-only fields are not modified, no saved variables are used, ...
7. Correctness proof of psyclone transforms? Can we show that our transforms do not create any bugs?

**Comment by arporter** [comment]
I'm thinking about @hiker's point 3. above - "Comparable debugging". This would also be useful when porting a code to e.g. GPU as one often needs to work out at which point in a run the results diverge. In theory, the PGI compiler (well, NVIDIA now) can do this at run-time by running a kernel on both CPU and GPU. However, I found that that made things even more unstable than they were already. Some compiler-independent functionality would be nice. We could do a 'truth' run which computed and saved (to disk) a checksum for each kernel executed. We'd then have a 'timeline' of checksums in a file. We'd then be able to repeat the run with the second executable (as long as the two runs used an identical series of kernels), read in the checksums, and compare them as we go until we find a divergence.

**Comment by arporter** [comment]
In fact, this timeline of kernels is something that has also been discussed as a requirement for assisting in building an adjoint of a model (required for Data Assimilation).

**Comment by hiker** [comment]
I am actually thinking of running the apps at the same time, and communicate the results at runtime (either using MPI dynamic process functionality (might not be supported on some platforms?), or sockets ... not trivial on some supercomputers, or ... something else ... like write-to-file *sigh* ... but at least that would be once per kernel only).
The problem when writing to disk (especially when doing full tests) is that the amount of space required can be quite prohibitive (though of course possible, e.g. for only a few kernels at a time).

**Comment by arporter** [comment]
Yes, I agree. It would be nice to run side-by-side but the complexity of doing that is going to be substantial. I was thinking of recording only kernel name and a checksum value so not very much data. Would still have to be careful with long runs though.

**Comment by hiker** [comment]
Ah yes, good point, that would work of course. Though direct communication means after a change is detected, we can send the 'correct' result to be used on the 'incorrect' side :P  (for the record, I already implemented this kind of library years ago ... but I couldn't open source it at that time). Hmm - as additional potential communication methods: a DB, kafka??

**Comment by rupertford** [comment]
Assuming we are not going to get the side-by-side working for a while, an alternative to the checksum-all-invokes option would be to gradually narrow down the region e.g. do per-timestep checks first, then major sections 2nd, then individual invokes 3rd, then kernels with full data comparisons. Of course, that would need us to know about the structure of the code somehow, probably via some infrastructure calls, which would allow us to selectively switch on checksums.

**Comment by hiker** [comment]
I have been thinking of some kind of ... tracing library. E.g. a runtime call-graph creation (which then could be used to visualise data dependency, e.g. you see data going into kernel 'a' is wrong, then you could 'somehow' use the call-graph data to see where 'a' was written to before ... quite vague obviously :)

**Comment by hiker** [comment]
As mentioned in the telco last night: I might be able to get a student to work on PSyclone for around 10 weeks. Given that it might require some initial training before this student can become productive, here a few short projects, from which the student could pick one or more. Note that suggestions assume that kernel extraction is working (since we are getting quite close, that seems like a valid assumption). For quite a few things we already have tickets.

- [x] Improve the output of the driver so that it produces statistics (atm it prints full fields when they are not bitwise identical, which is kind of useful). Better would be: X kernels are +-1%, y kernels +-5% or so
- [x] Support more than one input file for a kernel in the driver
- [x] Support MPI for driver (i.e. each process write its own files, which would then need 2. above to sequentially run a kernel with all input files).
- [x]  Create a testing environment that takes `X` driver automatically created, and `X` (or `n*X` if we have `n` output per kernel) data files, and provides a summary output in the end: "The following N kernels have bitwise identical results, ..., the following M kernels have difference of less than 0.1, ...
- [ ] I have a dummy 'tracing' library implemented: if you see a variable F having the wrong results as input (or output) to a kernel, you can lookup where this variable was written to before. The current implementation is based on the address of the variable/fields. Making this nice and shiny could be a useful tool

**Comment by arporter** [comment]
A lightweight debugging/profiling version of PSyData would be good - #2186. i.e. an automatic version of the most reliable and available debugging method aka `WRITE(*,*)`.

**Comment by arporter** [comment]
Thinking about @hiker's original point 5: "Creating unit tests for kernels? That might be reasonable easy with psyke(?)"
In the adjoint test-harness generation we construct an algorithm that initialises the necessary fields and then uses them to run a kernel (via an invoke). This is currently flagging some issues in the original kernels (as opposed to the adjointed ones).  It would therefore be good if we could generate such tests for *all* kernels (and build and run them in `full-debug`). This has the advantage that it is testing what the kernel metadata says the implementation expects. (If we use PSyKE, we are testing the kernel against inputs that the model is providing which might not match what the metadata claims is expected.)

We could also combine this test harness generation with the existing PSyclone functionality to generate code that checks that fields that are specified as READ only aren't modified by a kernel.

**Comment by hiker** [comment]
> A lightweight debugging/profiling version of PSyData would be good - #2186. i.e. an automatic version of the most reliable and available debugging method aka `WRITE(*,*)`.

Just printing the values of all input and output parameters is indeed a trivial exercise. It will just need a Fortran library to be linked in (with the kernel extraction transformation ... we might need to add the ability to specify if we want to include values used from other modules)

**Comment by hiker** [comment]
> Thinking about @hiker's original point 5: "Creating unit tests for kernels? That might be reasonable easy with psyke(?)" In the adjoint test-harness generation we construct an algorithm that initialises the necessary fields and then uses them to run a kernel (via an invoke). This is currently flagging some issues in the original kernels (as opposed to the adjointed ones). It would therefore be good if we could generate such tests for _all_ kernels (and build and run them in `full-debug`). This has the advantage that it is testing what the kernel metadata says the implementation expects. (If we use PSyKE, we are testing the kernel against inputs that the model is providing which might not match what the metadata claims is expected.)

Can you provide some additional details (maybe in the telco?)? We can likely do some static checking for gocean. Not sure about LFRic, due to its indirect addressing (and we can't assume that the content of these index array is correct?).

> We could also combine this test harness generation with the existing PSyclone functionality to generate code that checks that fields that are specified as READ only aren't modified by a kernel.

Yes, checksumming code is available in the psydata libs.

**Comment by arporter** [comment]
It would be good to do some experiments with applying PSyclone to WRF - it's widely used and has no official GPU port.

**Comment by arporter** [comment]
We have the opportunity to attempt to recruit a summer-placement student (undergraduate). I'm trying to think of (fun) things that would be suitable as a 6-12 week project. Some possibilities:
 * Online visualisation via PSyKE (@hiker has anyone done anything on this do you know?). We could perhaps aim to get a demo running on our Pi cluster (if it's still around) for e.g. NEMOLite2D.
 * Perhaps continuing @sergisiso 's implementation of a frontend using TreeSitter?
 * Is there any scope for working on/extending your Game of Life example @hiker?
 * Any other ideas?
I expect many people will have some Python experience but not many will have Fortran.

**Comment by hiker** [comment]
> We have the opportunity to attempt to recruit a summer-placement student (undergraduate). I'm trying to think of (fun) things that would be suitable as a 6-12 week project. Some possibilities:

Interesting!

> * Online visualisation via PSyKE ([@hiker](https://github.com/hiker) has anyone done anything on this do you know?). We could perhaps aim to get a demo running on our Pi cluster (if it's still around) for e.g. NEMOLite2D.

I have code that uses PSyData to go from Fortran via c to python, and can then visualise fields (e.g. while the app is running). That would need some cleanup (and optimisations, iirc there is a memcopy of the fields that I couldn't get immediately rid of, and that might be quite technical, i.e. mapping a Fortran field to numpy without copying), and it would need some additional work to get/handle coordinates, esp. if you want 3d visualisation.

Plus the main point: when I asked, no one was actually interested in in-situ visualisation ;) But it does make for a nice showcase (i.e. politicians visiting HPC centre can see moving important looking pictures :) ).

> * Perhaps continuing [@sergisiso](https://github.com/sergisiso) 's implementation of a frontend using TreeSitter?

Yes, faster parser would be nice. But tbh, I would estimate this being a lot of work. Getting the parser to work, and adding the conversion from parse tree to PSyIR? It would need a student with some background in this area?

> * Is there any scope for working on/extending your Game of Life example [@hiker](https://github.com/hiker)?

Hmm - I think I have taken this as far as I can ;)  The only small thing: a gui (which would tie in with the first point). But then again, most HPC experts are happy without moving pictures :)

Though, maybe I am missing things? Andy, you have read most of it. Any other ideas? Of course, they could do code-review/proof reading, but that doesn't fulfil the 'fun' criteria you mentioned ;)

> * Any other ideas?
- I intend to do some work in the automatic debugging area, but I think I am too far away to mentor, and I am pretty keen on that myself :)
- Doing the integration of Z3? We have a PR for this (and it's on my todo list to get back to that). Interesting, but tbh, I doubt that this is something for an undergrad (unless they have specialised on that). Actually, I think it's a stupid idea now, but I'll leave it here in case someone else has a good use case for adding more features if we have Z3 integrated.
- What about ML? We can use PSyData to feed fields for training/prediction? We already have an interface to python (see above).

>   I expect many people will have some Python experience but not many will have Fortran.
Yes, the usual problem ;)

**Comment by hiker** [comment]
Using Python as input language? I.e. be able to write kernels (and maybe alg layer??) in Python (with some restrictions of course), then create Fortran?

**Comment by arporter** [comment]
> Using Python as input language? I.e. be able to write kernels (and maybe alg layer??) in Python (with some restrictions of course), then create Fortran?

That sounds hard! @LonelyCat124 did a fair bit of work on a Particle DSL that had a python frontend though (IIRC).

**Comment by sergisiso** [comment]
> > Perhaps continuing [@sergisiso](https://github.com/sergisiso) 's implementation of a frontend using TreeSitter?

> Yes, faster parser would be nice. But tbh, I would estimate this being a lot of work. Getting the parser to work, and adding the conversion from parse tree to PSyIR? It would need a student with some background in this area?

I wouldn't discard this idea, I have the scaffolding necessary for parsing treesitter and creating CodeBlocks (and I can add UnknownFortranType) everywhere. The remaining can be added progressively in small steps. But it does require some interest in parsing which is not for everyone.

**Comment by LonelyCat124** [comment]
> > Using Python as input language? I.e. be able to write kernels (and maybe alg layer??) in Python (with some restrictions of course), then create Fortran?
> 
> That sounds hard! [@LonelyCat124](https://github.com/LonelyCat124) did a fair bit of work on a Particle DSL that had a python frontend though (IIRC).

I think for the DSLs this might be feasible for someone experienced with PSyclone/DSLs, but would not be feasible for a summer student I think - you'd need to learn python AST module usage, PSyclone PsyIR, how to interpose the two and how to fit them around the DSLs (and defining how the DSLs should look in Python).

Something to do with ML is probably interesting for a student, but I don't know if we have any use case for it at the moment w.r.t PSyclone?

**Comment by arporter** [comment]
ML might be interesting but we have zero knowledge and that would make supervision a bit tricky!

I hear what you say about moving pictures @hiker. A long time ago I was heavily involved in a project (https://github.com/RealityGrid/steer_lib) that enabled on-line vis. and the conclusion was that it was largely useful for demos to VIPs :-) (But also could be used for teaching students about how to use a model.)
